src/common/snippets/src/lowered/pass/pass.cpp

```diff
@@ -51,6 +51,26 @@ void PassPipeline::register_positioned_passes(const std::vector<PositionedPassLo
         register_pass(pp.position, pp.pass);
 }
 
+PassPipeline PassPipeline::merge_pipelines(const PassPipeline& lhs, const PassPipeline& rhs) {
+    OPENVINO_ASSERT(*lhs.get_pass_config() == *rhs.get_pass_config(), "2 passes with different PassConfigs can't be merged.");
+    const auto& lhs_passes = lhs.get_passes();
+    std::unordered_map<ov::DiscreteTypeInfo, std::shared_ptr<lowered::pass::PassBase>> passes_map;
+    for (const auto& pass : lhs_passes) {
+        passes_map[pass->get_type_info()] = pass;
+    }
+
+    auto merged_pipeline = lhs;
+    for (const auto& pass : rhs.get_passes()) {
+        auto lhs_pass_it = passes_map.find(pass->get_type_info());
+        if (lhs_pass_it == passes_map.end()) {
+            merged_pipeline.register_pass(pass);
+        } else {
+            OPENVINO_ASSERT(lhs_pass_it->second->can_be_merged(pass), "2 passes with type info ", pass->get_type_info(), " can't be merged.");
```

Comment:
Here we don't really perform transformation merge, do we?
Instead, we just check that if two transformations have similar `type_info` then they must have similar fields as well, so one of them could be skipped. So our `can_be_merged` is actually `is_same`.
The second concern here is that we check our `type_info` twice: first when we look it up in the map, and second when we try to cast it in `can_be_merged`. In other words, we already know that the pass has the same type info when we call `can_be_merged`, so there is no need to check it again.

As a possible solution, I think we can design a `PassPtr merge(const PassPtr&)` method that would actually perform the merge (which is very trivial in all present cases). If we also agree that an empty pass (aka nullptr) could be merged to any other pass, and if the merge fails then nullptr is returned. Then the merge would look like:
```suggestion
    for (const auto& pass : rhs.get_passes()) {
        auto& lhs_pass = passes_map[pass->get_type_info()];
        lhs_pass = pass->merge(lhs_pass);
        OPENVINO_ASSERT(lhs_pass, "2 passes with type info ", pass->get_type_info(), " can't be merged.");
        merged_pipeline.register_pass(lhs_pass);
    }
```
This would allow us to create a design that can handle complex handler merges

Reply:
Thanks for the proposal, I followed it. The only thing is that I had to add a little extra code (compared to your suggestion) to cover the case when `lhs` pipeline has more passes than `lhs`.